include/zephyr/drivers/misc/mcux_flexio/mcux_flexio.h

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2023, STRIM, ALC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_DRIVERS_MCUX_FLEXIO_H_
+#define ZEPHYR_DRIVERS_MCUX_FLEXIO_H_
+
+#include <zephyr/device.h>
+
+/**
+ * @struct mcux_flexio_child_res
+ * @brief Structure containing information about the required
+ * resources for a FlexIO child.
+ */
+struct mcux_flexio_child_res {
+	uint8_t *shifter_index;
+	uint8_t shifter_count;
+	uint8_t *timer_index;
+	uint8_t timer_count;
+};
+
+/**
+ * @typedef mcux_flexio_child_isr_t
+ * @brief Callback API to inform API user that FlexIO triggered interrupt
+ *
+ * This callback is called from IRQ context.
+ */
+typedef int (*mcux_flexio_child_isr_t)(void *user_data);
+
+/**
+ * @struct mcux_flexio_child
+ * @brief Structure containing the required child data for FlexIO
+ */
+struct mcux_flexio_child {
+	mcux_flexio_child_isr_t isr;
+	void *user_data;
+	struct mcux_flexio_child_res res;
+};
+
+/**
+ * @brief Enable FlexIO IRQ
+ * @param dev Pointer to the device structure for the FlexIO driver instance
+ */
+void mcux_flexio_irq_enable(const struct device *dev);
+
+/**
+ * @brief Disable FlexIO IRQ
+ * @param dev Pointer to the device structure for the FlexIO driver instance
+ */
+void mcux_flexio_irq_disable(const struct device *dev);
+
+/**
+ * @brief Lock FlexIO mutex.
+ * @param dev Pointer to the device structure for the FlexIO driver instance
+ */
+void mcux_flexio_lock(const struct device *dev);
```

Comment:
Just a note here- I'm not sure we should include `mcux_` as a prefix for these APIs. The driver does indeed rely on the MCUX HAL, but the API we are designing here is a consequence of the peripheral itself, not the HAL, and could probably be implemented without using the MCUX HAL if that was ever needed. Due to this, I think we should avoid the `mcux_` naming prefix here.

Reply:
@danieldegrasse, then it is also worth removing the `mcux_` prefix from the path (i.e. `include/zephyr/drivers/misc/flexio/flexio.h`) and adding `dev_api` to the mcux_flexio driver so that future flexio drivers can implement functions from this header.

Reply:
If you'd like to do that, I'm fine with both- I guess the reasoning being that another driver might exist implementing this API without the MCUX HAL? Truthfully I think if such a driver existed it would just replace the existing driver, so we might not need the ability for multiple drivers to exist. I am fine if you'd prefer to add the `dev_api` though.

Reply:
@danieldegrasse, flexio isn't really a driver in the Zephyr context, it's more of a driver for a driver that shouldn't provide access to users. So I suggest using `mcux_` prefixes in the file `include/zephyr/drivers/misc/flexio/flexio.h` and including this header in child drivers like this `#include "../misc/mcux_flexio/mcux_flexio.h"`

Reply:
I'm not sure I follow- will we have `mcux_` prefixes or not? I think users will absolutely access this driver, they need to do so in order to add support custom FLEXIO based peripheral drivers.

Reply:
MCUX sounds wrong as a prefix here, to be honest.

Reply:
I would strongly prefer we keep `mcux_` out of the naming scheme here. The API is not specific to MCUX SDK, it is enabling features of the FLEXIO IP. Can we place the file in `include/zephyr/drivers/misc/nxp_flexio/nxp_flexio.h`, and namespace functions with `nxp_flexio`?